toolchain/lexer/tokenized_buffer.cpp

```diff
@@ -48,6 +52,164 @@ auto VariantMatch(V&& v, Fs&&... fs) -> decltype(auto) {
+    }
+    for (char c = 'a'; c <= 'z'; ++c) {
+      table[c] = true;
+    }
+    table['_'] = true;
+    return table;
+  }();
+
+#if __x86_64__
+  // This code uses a scheme derived from the techniques in Geoff Langdale and
+  // Daniel Lemire's work on parsing JSON[1]. Specifically, that paper outlines
+  // a technique of using two 4-bit indexed in-register look-up tables (LUTs) to
+  // classify bytes in a branchless SIMD code sequence.
+  //
+  // [1]: https://arxiv.org/pdf/1902.08318.pdf
+  //
+  // The goal is to get a bit mask classifying different sets of bytes. For each
+  // input byte, we want the mask bits to be set with the following logic
+  // derived by inspecting the high nibble and low nibble of the input:
+  // bit0 = 1 for `_` needs 1 for high `0x5` and low `0xF`
+  // bit1 = 1 for `0-9` needs 1 for high `0x3` and low `0x0` - `0x9`
+  // bit2 = 1 for `A-O` needs 1 for high `0x4` and low `0x1` - `0xF`
+  // bit3 = 1 for `P-Z` needs 1 for high `0x5` and low `0x0` - `0xA`
+  // bit4 = 1 for `a-o` needs 1 for high `0x6` and low `0x1` - `0xF`
+  // bit5 = 1 for `p-z` needs 1 for high `0x7` and low `0x0` - `0xA`
+  // bit6 = unused
+  // bit7 = 1 for possible non-ASCII character in UTF-8, for all high nibbles
+  // with the high bit set and all low-nibbles.
+  //
+  // No bits set means definitively non-ID ASCII character.
+  //
+  // High bit set means contains the start of a non-ASCII character, need to
+  // fall back to a UTF-8 decoding path (not yet fully implemented).
+  //
+  // bit6 remains unused if we need to classify more characters.
+  const auto high_lut = _mm_setr_epi8(
+      /*0x0:*/ 0b0000'0000,
+      /*0x1:*/ 0b0000'0000,
+      /*0x2:*/ 0b0000'0000,
+      /*0x3:*/ 0b0000'0010,
+      /*0x4:*/ 0b0000'0100,
+      /*0x5:*/ 0b0000'1001,
+      /*0x6:*/ 0b0001'0000,
+      /*0x7:*/ 0b0010'0000,
+      /*0x8:*/ 0b1000'0000,
+      /*0x9:*/ 0b1000'0000,
+      /*0xA:*/ 0b1000'0000,
+      /*0xB:*/ 0b1000'0000,
+      /*0xC:*/ 0b1000'0000,
+      /*0xD:*/ 0b1000'0000,
+      /*0xE:*/ 0b1000'0000,
+      /*0xF:*/ 0b1000'0000);
+  const auto low_lut = _mm_setr_epi8(
+      /*0x0:*/ 0b1010'1010,
+      /*0x1:*/ 0b1011'1110,
+      /*0x2:*/ 0b1011'1110,
+      /*0x3:*/ 0b1011'1110,
+      /*0x4:*/ 0b1011'1110,
+      /*0x5:*/ 0b1011'1110,
+      /*0x6:*/ 0b1011'1110,
+      /*0x7:*/ 0b1011'1110,
+      /*0x8:*/ 0b1011'1110,
+      /*0x9:*/ 0b1011'1110,
+      /*0xA:*/ 0b1011'1100,
+      /*0xB:*/ 0b1001'0100,
+      /*0xC:*/ 0b1001'0100,
+      /*0xD:*/ 0b1001'0100,
+      /*0xE:*/ 0b1001'0100,
+      /*0xF:*/ 0b1001'0101);
+
+  int tail_utf8_mask;
+  int tail_ascii_mask;
+  int tail_mask;
+  // Use `ssize_t` for performance here as we index memory in a tight loop.
+  ssize_t i = 0;
+  const ssize_t size = text.size();
+  while ((i + 16) <= size) {
+    __m128i input =
+        _mm_loadu_si128(reinterpret_cast<const __m128i*>(text.data() + i));
+    __m128i low_mask = _mm_shuffle_epi8(low_lut, input);
```

Comment:
I think this doesn't quite work for the UTF-8 case: `_mm_shuffle_epi8` always gives `0` for a lane where the corresponding byte of `input` is negative, so we'll produce a `low_mask` of `0` for all non-ASCII bytes.

Fortunately, I think we can address this without extra masking here, by changing the computation of `tail_utf8_mask` below:

```diff
-    tail_utf8_mask = _mm_movemask_epi8(mask);
+    tail_utf8_mask = _mm_movemask_epi8(high_mask);
```

The point being, we don't need to look at the low-order nibble at all to classify UTF-8, so it doesn't matter that this won't compute it properly due to the behavior of `pshufb` for negative bytes.